### PR TITLE
Add conformance test for isolation of HTTP listeners

### LINKED ISSUE #2416: Listener isolation or cascading behavior
When talking about Listener acceptance and merging behavior as part of #2288, we discovered that there was a particularly important part of the spec that we had not clarified enough, and that this has created a behavior variance between implementations.

I (@youngnick) summarized this as follows:
> That concept is important because, for better or for worse, it has always been intended that the traffic-matching between Listeners and Routes is a two stage process. First you match a Listener, then you look for matching routes. In the case that there's a wildcard hostname, it was definitely not intended that a request could match both (for example, the `cafe.example.com` and `*.example.com` Listeners.) That is, having different values at all for hostname has been intended to make the Listeners distinct and unmergeable. We certainly have not done enough to explain that this is the intention, and I'm sorry for that.

That is, it was the original intention in the spec that, if there is a precise hostname match in one Listener, and a wildcard match in another, that traffic that matches the precise hostname will arrive _only at Routes bound to that Listener_.

Another way to put this is that if a precise hostname match is present, then a wildcard match should _never_ also match that precise hostname.

Because we were not clear about this in the spec originally, many implementations have implemented their rules such that requests that match a precise hostname may, if there are no matching Routes for the request, match Routes attached to the wildcard hostname.

To use an example, here is a Gateway that @pleshakov supplied in #2288.

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: gateway
spec:
  gatewayClassName: nginx
  listeners:
  - name: example
    port: 80
    protocol: HTTP
    hostname: "*.example.com"
  - name: cafe-http
    port: 80
    protocol: HTTP
    hostname: "cafe.example.com"
```

And here are some Routes that illustrate the config that _may_ result in cascading behavior. I've left out hostnames on the HTTPRoutes so that only the Listener hostname is relevant.

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: coffee
spec:
  parentRefs:
  - name: gateway
    sectionName: cafe-http
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /coffee
    backendRefs:
    - name: backend
      port: 3000
---
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: tea
spec:
  parentRefs:
  - name: gateway
    sectionName: example 
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /tea
    backendRefs:
    - name: backend 
      port: 3000
```

In this example, `cafe.example.com` has one HTTPRoute attached, `/coffee`, and `*.example.com` has one HTTPRoute attached, `/tea`.

The _original intent_ of the spec is that the only valid request using `cafe.example.com` is `http://cafe.example.com/coffee`. Any request that does not match the set of Routes attached to the `cafe-http` Listener should be 404ed.

However, many implementations allow the request to _cascade_ between Listeners, which means that request to either `http://cafe.example.com/coffee` _or_ `http://cafe.example.com/tea` will succeed.

This issue is to discuss how to proceed here.

My personal preference is to tighten up the language here, and make the originally-implied behavior explicit and compulsory, with conformance tests to back it up. I'm aware that this will be painful for a number of implementations, but I really think that this is better in the long run. This is a _very_ subtle difference that will be very difficult to explain to end users, and I think that it will significantly decrease portability to have this be optional in an extended feature.

### PR DESCRIPTION
**Pull request by pleshakov**
**What type of PR is this?**
/kind test
/area conformance

**What this PR does / why we need it**:

Listener isolation was clarified in https://github.com/kubernetes-sigs/gateway-api/pull/2465
This PR adds the corresponding extended feature and the conformance test.

The test includes two subtests:
- the one where hostnames are only configured in listeners
- the one where hostnames are configured both in listeners and HTTPRoutes, where HTTPRoutes try to "steal" requests bound to different listeners, but fail to do so because of the listener isolation.

Note:
I added only tests for HTTP listeners. I believe the fix for this issue https://github.com/kubernetes-sigs/gateway-api/issues/2417 might define the tests for the isolation for HTTPS listeners  - those will also involve setting SNI.

**Which issue(s) this PR fixes**:

Fixes #2416 

**Does this PR introduce a user-facing change?**:

```release-note
Adds an extended feature GatewayHTTPListenerIsolation with the corresponding conformance test GatewayHTTPListenerIsolation
```

### CODE REVIEW

**Comment by robscott** [comment]
Thanks @pleshakov! This is going to require some careful review from multiple people.

/cc @arkodg @mlavacca @sunjayBhatia 
/ok-to-test
/hold

**Review by sunjayBhatia — COMMENTED** [review]
ran this test real quick against Contour main and it does not pass, so I'll have to dig in further to see whats going on

```
--- FAIL: TestGatewayConformance (83.36s)
    --- FAIL: TestGatewayConformance/GatewayHTTPListenerIsolation (56.27s)
        --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners (0.03s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/4_request_to_'bar.example.com/empty-hostname'_should_receive_a_404 (0.01s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/12_request_to_'abc.foo.example.com/empty-hostname'_should_receive_a_404 (0.01s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/3_request_to_'bar.com/abc-foo-example-com'_should_receive_a_404 (0.02s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/2_request_to_'bar.com/wildcard-foo-example-com'_should_receive_a_404 (0.02s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/1_request_to_'bar.com/wildcard-example-com'_should_receive_a_404 (0.03s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/8_request_to_'bar.foo.example.com/empty-hostname'_should_receive_a_404 (0.03s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/11_request_to_'bar.foo.example.com/abc-foo-example-com'_should_receive_a_404 (0.03s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/7_request_to_'bar.example.com/abc-foo-example-com'_should_receive_a_404 (0.03s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/6_request_to_'bar.example.com/wildcard-foo-example-com'_should_receive_a_404 (0.04s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/0_request_to_'bar.com/empty-hostname'_should_go_to_infra-backend-v1 (0.04s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/10_request_to_'bar.foo.example.com/wildcard-foo-example-com'_should_go_to_infra-backend-v1 (0.03s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/9_request_to_'bar.foo.example.com/wildcard-example-com'_should_receive_a_404 (0.03s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/5_request_to_'bar.example.com/wildcard-example-com'_should_go_to_infra-backend-v1 (0.02s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/14_request_to_'abc.foo.example.com/wildcard-foo-example-com'_should_receive_a_404 (0.02s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/13_request_to_'abc.foo.example.com/wildcard-example-com'_should_receive_a_404 (0.02s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/hostnames_are_configured_only_in_listeners/15_request_to_'abc.foo.example.com/abc-foo-example-com'_should_go_to_infra-backend-v1 (0.02s)
        --- FAIL: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes (0.02s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/7_request_to_'bar.example.com/abc-foo-example-com'_should_receive_a_404 (0.01s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/0_request_to_'bar.com/empty-hostname'_should_go_to_infra-backend-v1 (0.01s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/5_request_to_'bar.example.com/wildcard-example-com'_should_go_to_infra-backend-v1 (0.01s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/15_request_to_'abc.foo.example.com/abc-foo-example-com'_should_go_to_infra-backend-v1 (0.02s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/6_request_to_'bar.example.com/wildcard-foo-example-com'_should_receive_a_404 (0.02s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/10_request_to_'bar.foo.example.com/wildcard-foo-example-com'_should_go_to_infra-backend-v1 (0.02s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/11_request_to_'bar.foo.example.com/abc-foo-example-com'_should_receive_a_404 (0.02s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/2_request_to_'bar.com/wildcard-foo-example-com'_should_receive_a_404 (0.01s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/3_request_to_'bar.com/abc-foo-example-com'_should_receive_a_404 (0.01s)
            --- PASS: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/1_request_to_'bar.com/wildcard-example-com'_should_receive_a_404 (0.01s)
            --- FAIL: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/14_request_to_'abc.foo.example.com/wildcard-foo-example-com'_should_receive_a_404 (30.00s)
            --- FAIL: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/4_request_to_'bar.example.com/empty-hostname'_should_receive_a_404 (30.00s)
            --- FAIL: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/13_request_to_'abc.foo.example.com/wildcard-example-com'_should_receive_a_404 (30.00s)
            --- FAIL: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/12_request_to_'abc.foo.example.com/empty-hostname'_should_receive_a_404 (30.00s)
            --- FAIL: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/8_request_to_'bar.foo.example.com/empty-hostname'_should_receive_a_404 (30.00s)
            --- FAIL: TestGatewayConformance/GatewayHTTPListenerIsolation/intersecting_hostnames_are_configured_in_listeners_and_HTTPRoutes/9_request_to_'bar.foo.example.com/wildcard-example-com'_should_receive_a_404 (30.00s)
FAIL
```

File: conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml
Comment (arkodg) [inline conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml:49]: these are more tied to hostname intersection b/w listener and route, should this be part of this test or another one ?

**Comment by arkodg** [comment]
thanks for adding this @pleshakov, your test approach looks good, added a minor comment

File: conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml
Comment (pleshakov) [reply conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml:49]: My interpretation is that listener intersection is about one listener and its HTTPRoutes (so it could be tested separately) -- https://github.com/kubernetes-sigs/gateway-api/blob/4b1a28d0c874a3816d344fb0d4e2c04424294d54/apis/v1/httproute_types.go#L71-L73

while listener isolation is about multiple listeners  -- https://github.com/kubernetes-sigs/gateway-api/blob/4b1a28d0c874a3816d344fb0d4e2c04424294d54/apis/v1/gateway_types.go#L173-L179

Or to put it differently, correct implementation of listener intersection is required for this test, but ultimately the tests is about listener isolation.

**Review by arkodg — APPROVED** [review]
/approve

thanks for adding this one !

**Comment by arkodg** [comment]
@sunjayBhatia do you want to take another look at this PR, or should we get this in ?

**Comment by robscott** [comment]
I've been meaning to take a look at this too, sorry I lost track of it! Will aim to get some feedback in this week.

/assign

**Comment by sunjayBhatia** [comment]
> @sunjayBhatia do you want to take another look at this PR, or should we get this in ?

yep I'll take another look tmr 👍🏽

**Review by sunjayBhatia — APPROVED** [review]
Contour needs an update to our Listener/Route hostname intersection logic to pass this but this looks like a nice thorough set of tests 👍🏽

File: conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml
Comment (robscott) [inline conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml:5]: Do we actually need 2 new Gateways for this test? (Asking because this could significantly increase the time it takes to run conformance tests + the overall resources required).

File: conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml
Comment (robscott) [inline conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml:1]:
```suggestion
apiVersion: gateway.networking.k8s.io/v1
```

**Review by robscott — COMMENTED** [review]
Thanks @pleshakov! This is an incredibly detailed test with well-written + explained test cases. For what it's worth, GKE Gateway ran into the same failures that @sunjayBhatia found, but that's on us, thanks for catching our inconsistencies with the spec! A few nits here + a rebase needed but otherwise LGTM.

File: conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml
Comment (pleshakov) [reply conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml:5]: > Do we actually need 2 new Gateways for this test? (Asking because this could significantly increase the time it takes to run conformance tests + the overall resources required).

No need for two Gateways, I can make it work with one to decrease time and required resources 👍

Comment (robscott) [reply conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml:5]: That sounds great, thanks! Ping me whenever this is ready for another look

**Review by mlavacca — APPROVED** [review]
Kong still needs to add logic to isolate HTTPS listeners, but the test looks well written and correct to me 👍

File: conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml
Comment (robscott) [reply conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml:5]: @pleshakov do you have time to merge these Gateways?

**Review by arkodg — APPROVED** [review]
/approve
/lgtm

File: conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml
Comment (pleshakov) [reply conformance/tests/gateway-http-listener-isolation-with-hostname-intersection.yaml:5]: Hi @robscott 

I didn't have enough time to dedicated to this. Apologies!

Is the idea to get the PR merged before the next Gateway API release?

One thing that worries me: 

since I already got approvals and confirmations from folks running this against their implementations, although merging should not cause any changes to the test behavior, there is a slight chance for it, since I will be updating code and manifests. So another round reviews and re-runs will be needed.

**Comment by arkodg** [comment]
@robscott I'd love to see this one into v1.1, this conformance test will benefit a lot of implementations and not supporting this is lapse in security, to make some progress, can we get this into v1.1 and add a follow up to optimize manifests / resources, this test is opt in, so shouldnt negatively affect anyone by default

**Comment by robscott** [comment]
@arkodg that makes sense to me, @pleshakov can you rebase this PR? RC2 is scheduled to go out tomorrow, if this PR is ready before that, I'll merge it.

**Comment by robscott** [comment]
Covered by #3047, closing this one out.

**Pull request closed without merging**
